Added alien path compatibility, Phi rotation and event pool randomisation to GenO2Kine

jackal1-66 opened the pull request:
This PR is related to the ticket https://its.cern.ch/jira/browse/O2-5216 
It has been tested locally and the feature works as expected. 
Check the ticket for more information.

jackal1-66 committed:
Implemented flag for .root kinematics input

jackal1-66 commented:
This PR is strictly connected to https://github.com/AliceO2Group/O2DPG/pull/1760

sawenzel commented on run/o2sim_parallel.cxx:
```diff
@@ -512,6 +512,46 @@ int main(int argc, char* argv[])
     return r;
   }
+  // check if Kinematics input file is provided
+  LOG(info) << "Kinematics input file: " << conf.getKineInput();
```
We have a generator called `GeneratorFromO2Kine`. This may already be doing most of the work... but we can extend it do be able to directly reading from Alien, applying some timeframe offsetting, and performing additional random modification (such as rotations).

jackal1-66 committed:
Event pooling with order randomisation of tree entries

jackal1-66 committed:
Added rng seed flag

jackal1-66 committed:
Added possibility of random rotation of Phi in event pool

jackal1-66 committed:
Added GRID kine input compatibility

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool1

jackal1-66 committed:
Introduction of GeneratorHybrid class for multiple generators configurations

At the moment the generation is random, but it's the starting point for event injection with external Kinematic files. The system has been tested locally and it seems to be working fine.

jackal1-66 committed:
Merge branch 'eventpool1' of https://github.com/jackal1-66/AliceO2 into eventpool1

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool1

jackal1-66 committed:
Implemented configurations and fractions parameters for hybrid generator

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool1

jackal1-66 committed:
Added features to use alien path, phi rotation and event randomisation

jackal1-66 committed:
Removal of unused implementation attempt

sawenzel commented on Generators/src/GeneratorFromFile.cxx:
```diff
@@ -29,6 +29,13 @@ namespace eventgen
 {
 GeneratorFromFile::GeneratorFromFile(const char* name)
 {
+  if (strncmp(name, "alien:/", 7) == 0) {
```
I think GeneratorFromFile is meant to read from Run2 AliRoot kinematics files. Probably the changes here do not make sense for these files (we can concentrate on the GeneratorFromO2Kine case).

jackal1-66 replied:
Done

jackal1-66 committed:
Deleted feature for RUN2 based generator

sawenzel commented:
Can you also change the title of the PR and remove the WIP?

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool

jackal1-66 committed:
Merge branch 'AliceO2Group:dev' into eventpool1

jackal1-66 committed:
Merge branch 'eventpool1' into eventpool

jackal1-66 committed:
Revert "Add SMatrixGPU compatibility to trackParCov (#13602)"

This reverts commit 6a2cc7e940863599154897007ac556cb51e72f59.

jackal1-66 committed:
Revert "Merge branch 'eventpool1' into eventpool"

This reverts commit 2d3b69218d6f235f2a63846d5025df171bf92503, reversing
changes made to c9a5dad46522d515a7bdf50a3a992dc3f123c90f.

jackal1-66 commented:
There's a bit of a mess with the commits. Please DO NOT MERGE until I've sorted them out.

jackal1-66 committed:
Reapply "Add SMatrixGPU compatibility to trackParCov (#13602)"

This reverts commit 335f4c9eb5f1aefdf2aec18804f80d746c939c5e.

jackal1-66 commented:
Problem fixed. The merging can go forward.

sawenzel commented on Generators/src/GeneratorFromFile.cxx:
```diff
@@ -253,6 +278,13 @@ bool GeneratorFromO2Kine::importParticles()
+        phi += dPhi;
+        px = pt * TMath::Cos(phi);
+        py = pt * TMath::Sin(phi);
+      }
```
I believe this can be done with less cost by applying the rotation as a matrix - vector multiplication:

```
px =  cos phi * px_old + sin (-phi) * py_old 
py =  sin phi * px_old + cos (phi) * py_old
```
This would only need evaluation of cos, sin and not Atan2 nor sqrt.

jackal1-66 committed:
Better rotation handling
